### PR TITLE
Closes #6460: Upstream RunWhenReadyQueue

### LINKED ISSUE #6460: Move RunWhenReadyQueue into a-c
https://github.com/mozilla-mobile/fenix/pull/9239 adds `RunWhenReadyQueue` which could be useful in a few places; let's move it into a-c, and add some tests for it.

### PR DESCRIPTION
**Pull request by NotWoods**
---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/support/base/src/main/java/mozilla/components/support/base/concurrent/RunWhenReadyQueue.kt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+     * Runs the [task] if this queue is marked as ready, or queues it for later execution.
+     *
+     * @param task: The task to run now if queue is ready or queue for later execution.
+     */
+    fun runIfReadyOrQueue(task: () -> Unit) {
+        if (isReady.get()) {
+            scope.launch { task.invoke() }
+        } else {
+            tasks.add(task)
+        }
```
Comment (pocmo) [inline components/support/base/src/main/java/mozilla/components/support/base/concurrent/RunWhenReadyQueue.kt:42]: This part doesn't seem thread-safe is `runIfReadyOrQueue()` and `ready()` are called on different threads? I think we would need to synchronize here? Or alternatively assert that we are on the main thread if that is what we expect.

Comment (NotWoods) [reply components/support/base/src/main/java/mozilla/components/support/base/concurrent/RunWhenReadyQueue.kt:42]: Would `AtomicBoolean` handle the threads? cc @grigoryk who worked on the class

Comment (grigoryk) [reply components/support/base/src/main/java/mozilla/components/support/base/concurrent/RunWhenReadyQueue.kt:42]: I think what _could_ happen here is that `tasks` is modified by `runIfReadyOrQueue` while we're iterating through it in `ready`. This won't result in an exception thanks to `CopyOnWriteArrayList`, but we will drop that task on the floor in the `tasks.clear` call right after the iteration, I think.

A `synchronized` block wrapping the iteration/clearing of `tasks`, and wrapping `tasks.add` should be fine, though. With that in place, you can switch to a regular array, as well.

**Pull request merged**